src/hash_table_internal.h

```diff
@@ -46,32 +43,67 @@
  */
 struct ly_ht_rec {
     uint32_t hash;        /* hash of the value */
-    int32_t hits;         /* collision/overflow value count - 1 (a filled entry has 1 hit,
-                           * special value -1 means a deleted record) */
-    unsigned char val[1]; /* arbitrary-size value */
-} _PACKED;
```

Comment:
What is the problem with having the records packed? You have encountered some alignment issues? Because it can save a lot of memory.

Reply:
I did not encounter any alignment issue.

I think it was correct before, but changing the size of ly_ht_rec makes it incorrect. Let me try to explain the idea.

To me, the packed attribute imply 2 things:
1/ no padding is added between fields. In this case, we only have u32, so there is no padding even without the attribute.
2/ the compiler will generate a (slower) code to access the structure fields that will work whatever the alignment of the structure. This will impact access to ly_ht_rec structure, but not structure located right after (which have probably no packed attribute).

Before my patches, sizeof(struct ly_ht_rec) is 8, so we don't really change the alignment constraint of the structure located right after. So, the accesses to the record always match the default alignment, and accesses to struct ly_ht_rec can be anywhere thanks to the packed attribute.

But the 2nd patch of my patchset changes the size of struct ly_ht_rec from 8 to 12. If alignment constraint of the record is 8, it could trigger unaligned accesses on some architectures.

So the idea was to ensure that size(ly_ht_rec) + record_size is always aligned.

I think the best solution would be to understand how to remove the 2nd patch without breaking the tests, to keep the packed attribute as before, and to add a compile-time check to ensure that size(struct ly_ht_rec) == 8.

What do you think?

Reply:
To me, it mainly implies that the records will not waste (possibly a lot of) space, for huge hash tables. But it is true that it is used only for aligned values (usually pointers) so it makes no difference. However, I seem to remember that the actual reason I have added that attribute was to fix endianity problems. Because if the value is padded to be larger, assigning it may use the padded bytes and when reading it back, they may be ignored, something along these lines. But again the same argument, it worked until now, why change it?

Reply:
> But again the same argument, it worked until now, why change it?

The patchset introduces a change: before, if a record had an alignment constraint of 8, it was ensured that the start of the record structure was aligned to 8. If I keep the packed attribute as is, the record will be aligned to 4.

It is probably not a problem on modern archs, but I had the feeling that we should avoid this (see https://www.kernel.org/doc/html/next/core-api/unaligned-memory-access.html).

Anyway, I'll remove this change in v2, we can still come back on it if needed.

Thanks

Reply:
Maybe I am missing something but I do not think it was (is currently) aligned to anything. If a hash table is created with value size 1 B, there will be no alignment, every record would have 9 B (the real size since it is packed) and the records are one after the other in an array. So technically, the alignment is the value size. But you have not changed the size of the record so like I said, it should work just as before.

Reply:
Yes, for a structure of size 1B has an alignment constraint of 1B, if the sum has a size of 9B, it is still ok.

But if we take a structure of which needs to be aligned to 8B:
struct foo {
  uint64_t x;
  char y;
};

Its size is 16B because the compiler adds padding, to keep all elements aligned when doing an array of struct foo.

The sum size(ly_ht_rec) + record_size was 24B before my patches (so, still ok), but it is 28B after, which breaks the alignment constraint.

Reply:
> Its size is 16B because the compiler adds padding, to keep all elements aligned when doing an array of struct foo.

Well, to be precise, this is not always true, it depends on ABI.

Reply:
Okay, I have never put structures into the hash table but yes, if the structure is not packed, `sizeof` should return its padded size. But we are not doing this internally, like I said, so I would prefer to keep the records packed to save some memory.

Reply:
Got it, will do like this in v2, thanks!